plugins/georeplication/rest.go

```diff
@@ -0,0 +1,193 @@
+		return nil, nil, errs.New("Invalid Master Volume ID")
+	}
+
+	// Validate UUID format of Slave Volume ID
+	slaveid := uuid.Parse(slaveidRaw)
+	if slaveid == nil {
+		restutils.SendHTTPError(w, http.StatusBadRequest, "Invalid Slave Volume ID")
+		return nil, nil, errs.New("Invalid Slave Volume ID")
+	}
+
+	return masterid, slaveid, nil
+}
+
+func georepCreateUpdateHandler(w http.ResponseWriter, r *http.Request) {
+	// Collect inputs from URL
+	p := mux.Vars(r)
+	masteridRaw := p["mastervolid"]
+	slaveidRaw := p["slavevolid"]
+
+	reqID, logger := restutils.GetReqIDandLogger(r)
+
+	// Same handler for both Create and Update, if HTTP method is POST then
+	// it is update request
+	updateRequest := false
+	if r.Method == http.MethodPost {
+		updateRequest = true
+	}
+
+	// Validate UUID format of Master and Slave Volume ID
+	masterid, slaveid, err := validateMasterAndSlaveIDFormat(w, masteridRaw, slaveidRaw)
+	if err != nil {
+		return
+	}
+
+	// Parse the JSON body to get additional details of request
+	var req georepapi.GeorepCreateRequest
+	if err := utils.GetJSONFromRequest(r, &req); err != nil {
+		restutils.SendHTTPError(w, http.StatusUnprocessableEntity, errors.ErrJSONParsingFailed.Error())
+		return
+	}
+
+	// Required fields are MasterVol, SlaveHosts and SlaveVol
+	if req.MasterVol == "" {
+		restutils.SendHTTPError(w, http.StatusBadRequest, "Master volume name is required field")
+		return
+	}
+
+	if len(req.SlaveHosts) == 0 {
+		restutils.SendHTTPError(w, http.StatusBadRequest, "Atleast one Slave host is required")
+		return
+	}
+
+	if req.SlaveVol == "" {
+		restutils.SendHTTPError(w, http.StatusBadRequest, "Slave volume name is required field")
+		return
+	}
+
+	// Check if Master volume exists and Matches with passed Volume ID
+	vol, e := volume.GetVolume(req.MasterVol)
+	if e != nil {
+		restutils.SendHTTPError(w, http.StatusNotFound, errors.ErrVolNotFound.Error())
+		return
+	}
+
+	// Check if Master Volume ID from store matches the input Master Volume ID
+	if !uuid.Equal(vol.ID, masterid) {
+		restutils.SendHTTPError(w, http.StatusBadRequest, "Master volume ID doesn't match")
+		return
+	}
+
+	// Fetch existing session details from Store, if same
+	// session exists then return error
+	geoSession, err := getSession(masterid.String(), slaveid.String())
+	if err == nil {
+		if !updateRequest {
+			restutils.SendHTTPError(w, http.StatusConflict, "Session already exists")
+			return
+		}
+	} else {
+		// Continue only if NotFound error, return if other errors like
+		// error while fetching from store or JSON marshal errors
+		if _, ok := err.(*ErrGeorepSessionNotFound); !ok {
+			restutils.SendHTTPError(w, http.StatusInternalServerError, err.Error())
+			return
+		}
+		// Fail if update request received for non existing Geo-rep session
+		if updateRequest {
+			restutils.SendHTTPError(w, http.StatusBadRequest, "geo-replication session not found")
+			return
+		}
+
+		geoSession = new(georepapi.GeorepSession)
```

Comment:
Consider populating this struct instance (including members filled below) in it's own function, may be something like:

```go
func newGeorepSession(req *georepapi.GeorepCreateRequest) *georepapi.GeorepSession
```

Reply:
Ok